src/os/bluestore/BlueFS.cc

```diff
@@ -2229,6 +2258,177 @@ int64_t BlueFS::_read_random(
+  uint64_t wal_data_logical_offset = 0;
+
+
+  // save previous position as buffer pos is treated difffernt on regular files
+  uint64_t previous_pos = h->buf.pos;
+
+  uint64_t remaining_len = len;
+  auto flush_iterator = h->file->wal_flushes.begin();
+  while (remaining_len > 0 && flush_iterator != h->file->wal_flushes.end()) {
+    uint64_t flush_offset = flush_iterator->offset;
+    uint64_t flush_length = flush_iterator->length;
+    dout(20) << __func__ << " flush_offset " << flush_offset << dendl;
+    dout(20) << __func__ << " flush_length " << flush_length << dendl;
+
+    if (flush_length == 0) {
+      if (remaining_len > 0) {
+        dout(5) << __func__ << " flush_length 0: reading less then required "
+          << ret << "<" << len - ret << dendl;
+      }
+      break;
+    }
+    // if we won't find offset here, go ahead
+    bool in_range = wal_data_logical_offset < off+len && wal_data_logical_offset+flush_length > off;
+    if (!in_range) {
+      if (off >= wal_data_logical_offset + flush_length) {
+        // move to next flush
+        // TODO(pere): do we check "ino" here too?
+        wal_data_logical_offset += flush_length;
+        flush_iterator++;
+        continue;
+      }
+      // we can't read more, should this happen?
+      dout(5) << __func__ << " unexpected not in range: reading less then required "
+        << ret << "<" << len - ret << dendl;
+      break;
+
+    }
+
+    flush_offset += sizeof(File::WALFlush::WALLength);
+
+    uint64_t skip_front = 0;
+    if(wal_data_logical_offset < off) { 
+      // offset is in this flush chunk so if we are before we move forward
+      skip_front = off - wal_data_logical_offset;
+    }
+    flush_offset += skip_front;
+    wal_data_logical_offset += skip_front;
+
+    dout(20) << __func__ << " after getting flush flush_offset " << flush_offset 
+      << " skipping: " << skip_front << dendl;
+
+    uint64_t data_to_read_from_flush = std::min(flush_length-skip_front, remaining_len);
+    bufferlist payload;
+    dout(20) << __func__ << "  data_to_read_from_flush" << data_to_read_from_flush << dendl;
+    _read(h, flush_offset, data_to_read_from_flush, &payload, nullptr);
+
+    if (outbl) {
+      outbl->append(payload);
+    }
+    if (out) {
+      auto p = payload.begin();
+      p.copy(data_to_read_from_flush, out);
+      out += data_to_read_from_flush;
+    }
+
+    // Check marker
+    {
+      uint64_t marker_offset = flush_iterator->get_marker_offset();
+      bufferlist marker_bl;
+      _read(h, marker_offset, sizeof(File::WALFlush::WALMarker), &marker_bl, nullptr);
+      uint64_t marker = *((uint64_t*)marker_bl.c_str());
+      dout(20) << "marker " << marker << " " << h->file->fnode.ino << dendl;
+      ceph_assert(marker == h->file->fnode.ino);
+    }
+    flush_iterator++;
+
+    remaining_len -= data_to_read_from_flush;
+    wal_data_logical_offset += data_to_read_from_flush;
+    ret += data_to_read_from_flush;
+  }
+  if (remaining_len > 0) {
```

Comment:
IMO better print remaining_len in the summary logging below (at line 2424). I see no much sense in printing this "warning" as a standalone line using the same debug level...